### ISSUE TITLE
The XLIB crash bug on Debian

### ISSUE THREAD

**Issue by cebe** [uTox/uTox#742]
Here is what I think does happen:

The story of crashing xlib on debian with https://github.com/uTox/uTox/pull/550

1. [main()](https://github.com/GrayHatter/uTox/blob/68df24b4501026174f037d5ec7429d2a1415cf5e/src/xlib/main.c#L641) it begins...
2. [xlib gets initialized a bit](https://github.com/GrayHatter/uTox/blob/68df24b4501026174f037d5ec7429d2a1415cf5e/src/xlib/main.c#L678-L700)
3. [thread(toxcore_thread, NULL);](https://github.com/GrayHatter/uTox/blob/68df24b4501026174f037d5ec7429d2a1415cf5e/src/xlib/main.c#L702)

From here on there are two threads:

| main  | toxcore_thread
| ----  | ----- 
| [initialize main_window.drawbuf](https://github.com/GrayHatter/uTox/blob/68df24b4501026174f037d5ec7429d2a1415cf5e/src/xlib/main.c#L713)       | [init toxcore and call init_self()](https://github.com/GrayHatter/uTox/blob/68df24b4501026174f037d5ec7429d2a1415cf5e/src/tox.c#L444-L477)
|  more     | [init_self() calls avatar_init_self()](https://github.com/GrayHatter/uTox/blob/68df24b4501026174f037d5ec7429d2a1415cf5e/src/self.c#L28)
|  xlib calls     | [avatar_init_self() calls avatar_load()](https://github.com/GrayHatter/uTox/blob/68df24b4501026174f037d5ec7429d2a1415cf5e/src/avatar.c#L175)
|  going here     | [avatar_load() calls utox_image_to_native()](https://github.com/GrayHatter/uTox/blob/68df24b4501026174f037d5ec7429d2a1415cf5e/src/avatar.c#L97)
|  ...     | [utox_image_to_native() calls XCreateImage()](https://github.com/GrayHatter/uTox/blob/68df24b4501026174f037d5ec7429d2a1415cf5e/src/xlib/main.c#L453) using global variables `display` and `default_visual` which are constantly used in method calls on the main thread while that one is further initializing the window and drawing stuff.

The above rows may race and cause the xserver to hang, making the system unresponsive until I kill `utox`.

**What is wrong?**

Even though we call [`XInitThreads()`](https://anonscm.debian.org/cgit/pkg-xorg/lib/libx11.git/tree/src/locking.c?h=libx11-2_1.6.2-3#n572) which initializes some locks, there seems to be a bug in xlib that can not deal with
how utox is calling it.

[`XWhitePixel` and `XBlackPixel`](https://anonscm.debian.org/cgit/pkg-xorg/lib/libx11.git/tree/src/Macros.c?h=libx11-2_1.6.2-3#n64) seem to be accessing the display without locking
[as far as I see](https://anonscm.debian.org/cgit/pkg-xorg/lib/libx11.git/tree/include/X11/Xlib.h?h=libx11-2_1.6.2-3#n135) not sure if that is a problem though.

[XCreateIC](https://anonscm.debian.org/cgit/pkg-xorg/lib/libx11.git/tree/src/xlibi18n/ICWrap.c?h=libx11-2_1.6.2-3#n226) is calling some method involving its arguments without locking.
And we are passing the main_window [here](https://github.com/GrayHatter/uTox/blob/68df24b4501026174f037d5ec7429d2a1415cf5e/src/xlib/main.c#L778-L779).
After that there is also the tray icon creation, which I have not checked in more detail because the bug is also reproduceable with tray icon commented out.

Links above reference the xlib code of the exact version I have installed: [libx11-2_1.6.2-3](https://anonscm.debian.org/cgit/pkg-xorg/lib/libx11.git/tree/?h=libx11-2_1.6.2-3).

**How to fix?**

1. In general the toxcore thread should not call anything avatar related at all.
2. but to make the fix easier, we could move the initialization of the toxcore thread to a later point at which the window is already created
   and the avatar related data structures are fully initialized.

Here is a patch:

```diff
diff --git a/src/xlib/main.c b/src/xlib/main.c
index b0b17d7..88c3e88 100644
--- a/src/xlib/main.c
+++ b/src/xlib/main.c
@@ -699,7 +699,6 @@ int main(int argc, char *argv[]) {
 
     atom_init();
 
-    thread(toxcore_thread, NULL);
 
     LANG = systemlang();
     dropdown_language.selected = dropdown_language.over = LANG;
@@ -801,6 +800,9 @@ int main(int argc, char *argv[]) {
     native_window_set_target(&main_window);
     panel_draw(&panel_root, 0, 0, settings.window_width, settings.window_height);
 
+    // start toxcore thread
+    thread(toxcore_thread, NULL);
+
     /* event loop */
     while (true) {
         XEvent event;
```

**Comment by cebe** [comment]
I would have liked to find the bug in xlib and fix that too but I did not see any suspicious code in there.

**Comment by cebe** [comment]
fixed by GrayHatter/uTox#3  in #550.
